File: Selection/SelectionWorst.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <algorithm>

using namespace std;

int findMedian(vector<int> &arr, int start, int n)
{
    sort(arr.begin() + start, arr.begin() + start + n);
    return arr[start + n / 2];
}

int partition(vector<int> &arr, int low, int high, int pivot)
{
    int i;
    for (i = low; i < high; ++i)
    {
        if (arr[i] == pivot)
            break;
    }
    swap(arr[i], arr[high]);

    i = low;
    for (int j = low; j < high; ++j)
    {
        if (arr[j] <= pivot)
        {
            swap(arr[i], arr[j]);
            i++;
        }
    }
    swap(arr[i], arr[high]);

    return i;
}

int select(vector<int> &arr, int low, int high, int k)
{
    if (low == high)
        return arr[low];

    int n = high - low + 1;
    vector<int> medians((n + 4) / 5);

    for (int i = 0; i < n / 5; ++i)
        medians[i] = findMedian(arr, low + i * 5, 5);
    if (n % 5)
        medians[n / 5] = findMedian(arr, low + (n / 5) * 5, n % 5);

    int medOfMed = (medians.size() == 1) ? medians[0]
                                         : select(medians, 0, medians.size() - 1, medians.size() / 2);

    int pos = partition(arr, low, high, medOfMed);
    int order = pos - low + 1;

    if (order == k)
        return arr[pos];
    else if (k < order)
        return select(arr, low, pos - 1, k);
    else
        return select(arr, pos + 1, high, k - order);
}

int main(int argc, char *argv[])
{
    if (argc < 3)
    {
        cerr << "Usage: ./SelectionWorst <input.txt> <i-th>\n";
        return 1;
    }

    string inputFile = argv[1];
    int k = stoi(argv[2]);

    ifstream infile(inputFile);
    if (!infile)
    {
        cerr << "Failed to open " << inputFile << '\n';
        return 1;
    }

    vector<int> arr;
    int x;
    while (infile >> x)
        arr.push_back(x);
    infile.close();

    if (k < 1 || k > arr.size())
    {
        cerr << "Error: i-th number out of range.\n";
        return 1;
    }

    int minVal = *min_element(arr.begin(), arr.end());
    int maxVal = *max_element(arr.begin(), arr.end());
    int kthElement = select(arr, 0, arr.size() - 1, k);

    cout << "Min = " << minVal << " Max = " << maxVal << endl;
    cout << "The " << k << "-th = " << kthElement << endl;

    return 0;
}

File: Selection/SelectionWorst.py
import sys


def find_median(arr, start, n):
    arr[start:start + n] = sorted(arr[start:start + n])
    return arr[start + n // 2]


def partition(arr, low, high, pivot):
    i = next((idx for idx in range(low, high) if arr[idx] == pivot), high)
    arr[i], arr[high] = arr[high], arr[i]

    i = low
    for j in range(low, high):
        if arr[j] <= pivot:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
    arr[i], arr[high] = arr[high], arr[i]

    return i


def select(arr, low, high, k):
    if low == high:
        return arr[low]

    n = high - low + 1
    medians = [
        find_median(arr, low + i * 5, min(5, n - i * 5))
        for i in range((n + 4) // 5)
    ]

    med_of_med = select(medians, 0,
                        len(medians) - 1,
                        len(medians) // 2) if len(medians) > 1 else medians[0]

    pos = partition(arr, low, high, med_of_med)
    order = pos - low + 1

    if order == k:
        return arr[pos]
    elif k < order:
        return select(arr, low, pos - 1, k)
    else:
        return select(arr, pos + 1, high, k - order)


def main():
    if len(sys.argv) < 3:
        print("Usage: python SelectionWorst.py <input.txt> <i-th>")
        return

    input_file, k = sys.argv[1], int(sys.argv[2])

    try:
        with open(input_file, "r") as f:
            arr = list(map(int, f.read().strip().split()))
    except FileNotFoundError:
        print(f"Failed to open {input_file}")
        return

    if k < 1 or k > len(arr):
        print("Error: i-th number out of range.")
        return

    min_val = min(arr)
    max_val = max(arr)
    kth_element = select(arr, 0, len(arr) - 1, k)

    print(f"Min = {min_val} Max = {max_val}")
    print(f"The {k}-th = {kth_element}")


if __name__ == "__main__":
    main()
